LRGB: Long Range Graph Benchmark datasets

vijaydwivedi75 opened the pull request:
This PR adds the LRGB datasets from the paper [Long Range Graph Benchmark](https://openreview.net/pdf?id=in7XC5RcjEn). The original dataset source is [in this repo](http://github.com/vijaydwivedi75/lrgb).

The Long Range Graph Benchmark (LRGB) is a collection of 5 graph learning datasets that arguably require long-range reasoning to achieve strong performance in a given task. The 5 datasets in this benchmark can be used to prototype new models that can capture long range dependencies in graphs.

Dataset | Domain | Task
-- | -- | --
PascalVOC-SP | Computer Vision | Node Classification
COCO-SP | Computer Vision | Node Classification
PCQM-Contact | Quantum Chemistry | Link Prediction
Peptides-func | Chemistry | Graph Classification
Peptides-struct | Chemistry | Graph Regression

The `torch_geometric.datasets.LRGBDataset` can be used to access any of the 5 datasets in the benchmark.

EdisonLeeeee commented:
Thanks for adding these! Left some initial comments.

EdisonLeeeee commented on torch_geometric/datasets/lrgb.py:
```diff
@@ -0,0 +1,330 @@
+                  self.raw_dir)
+        os.unlink(path)
+
+    def process(self):
```
Since `process_sp`  and `process_peptides` work in a similar way, can we unify them into one single processing function?

EdisonLeeeee commented on torch_geometric/datasets/lrgb.py:
```diff
@@ -0,0 +1,330 @@
+                if self.name == 'coco-sp':
+                    # Label remapping for coco-sp.
+                    # See self.label_remap_coco() func
+                    label_map = self.label_remap_coco()
```
This can be simplified with `sklearning.preprocessing.LabelEncoder`, e.g., 
```python
from sklearn.preprocessing import LabelEncoder
y = LabelEncoder().fit_transform(graph[3])
y = torch.LongTensor(y)
```

EdisonLeeeee commented on torch_geometric/datasets/lrgb.py:
```diff
@@ -0,0 +1,330 @@
+            :header-rows: 1
+
+            * - Name
+              - Task
```
An additional column `Task` seems not supported in the `CheatSheet` part

vijaydwivedi75 commented:
Thanks @EdisonLeeeee for your comments! 
I have included the following revisions in an updated commit.
- Merged process steps for the superpixels and Peptides datasets
- Removed 'Task' column from the Stats
- Simplified contiguous label through LabelEncoder.

EdisonLeeeee commented on torch_geometric/datasets/lrgb.py:
```diff
@@ -0,0 +1,293 @@
+                pbar.set_description(f'Processing {split} dataset')
+
+                data_list = []
+                for idx in indices:
```
```suggestion
                for graph in tqdm(graphs, desc=f'Processing {split} dataset'):
```

EdisonLeeeee commented on torch_geometric/datasets/lrgb.py:
```diff
@@ -0,0 +1,293 @@
+                edge_index = graph[2]
+                edge_index_labeled = graph[3]
+                edge_label = graph[4]
+                y = None
```
```suggestion
```

EdisonLeeeee commented on torch_geometric/datasets/lrgb.py:
```diff
@@ -0,0 +1,293 @@
+
+                data = Data(x=x, edge_index=edge_index, edge_attr=edge_attr,
+                            edge_index_labeled=edge_index_labeled,
+                            edge_label=edge_label, y=y)
```
```suggestion
                            edge_label=edge_label)
```

EdisonLeeeee commented on torch_geometric/datasets/lrgb.py:
```diff
@@ -0,0 +1,293 @@
+                        edge_attr = graph[1].to(torch.float)
+                        edge_index = graph[2]
+                        if self.name == 'coco-sp':
+                            y = label_map_coco.transform(graph[3])
```
I think we can directly use `LabelEncoder.fit_transform(graph[3])` here without the need of `original_label_idxs`, wdyt?

vijaydwivedi75 replied:
This is because one graph sample will only have a subset of the labels present in `original_label_idxs`, and hence would require to fit on the full label set.

EdisonLeeeee replied:
Oh, I see. You are right!

vijaydwivedi75 commented on torch_geometric/datasets/lrgb.py:
```diff
@@ -0,0 +1,293 @@
+from typing import Callable, List, Optional
+
+import torch
+from sklearn.preprocessing import LabelEncoder
```
If sklearn is not there by default and the `original_label_idxs` is required, should I revert back to using the `label_remap()` function that was in the first version?

EdisonLeeeee replied:
I have two suggestions:
+ Import `sklearn` internally as https://github.com/pyg-team/pytorch_geometric/blob/062556307bf6ebca3f0f185db8fd2dafd7362313/torch_geometric/nn/models/node2vec.py#L173
+ Implement `LabelEncoder` simply with PyTorch. FYI:
```python
label_encoder = torch.zeros(max(original_label_idxs))
label_encoder[original_label_idxs] = torch.arange(len(original_label_idxs))
...

y = label_encoder[torch.LongTensor[graphs[2]])
```

vijaydwivedi75 replied:
I feel we can skip the usage of sklearn as (i) its a one-time thing for `COCO-SP` and does not make much difference per my runs, and (ii) we require the label set `original_label_idxs`.

EdisonLeeeee replied:
Agreed. Not a strong preference for the use of sklearn.

vijaydwivedi75 commented:
Hi @EdisonLeeeee I have updated the code based on the feedbacks -thank you! 
It is ready from my side.

EdisonLeeeee commented on torch_geometric/datasets/lrgb.py:
```diff
@@ -0,0 +1,278 @@
+                    if self.name == 'coco-sp':
+                        # Label remapping for coco-sp.
+                        # See self.label_remap_coco() func
+                        label_map = self.label_remap_coco()
```
We are not necessarily to call this method and get `label_map` at each iteration, right?

vijaydwivedi75 replied:
Right -thanks! It can be called just once for coco. I updated the file.

EdisonLeeeee commented on CHANGELOG.md:
```diff
@@ -5,6 +5,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ## [2.2.0] - 2022-MM-DD
 ### Added
+- Added `LRGBDataset` to include 5 datasets from the [Long Range Graph Benchmark](https://openreview.net/pdf?id=in7XC5RcjEn), [#5935](https://github.com/pyg-team/pytorch_geometric/pull/5935)
```
```suggestion
- Added `LRGBDataset` to include 5 datasets from the [Long Range Graph Benchmark](https://openreview.net/pdf?id=in7XC5RcjEn) ([#5935](https://github.com/pyg-team/pytorch_geometric/pull/5935))
```

EdisonLeeeee commented on torch_geometric/datasets/lrgb.py:
```diff
@@ -0,0 +1,280 @@
+                x = graph[0]
+                edge_attr = graph[1]
+                edge_index = graph[2]
+                edge_index_labeled = graph[3]
```
[naming] Use `edge_label_index` to be consistent with data split by `torch_geometric.transforms.RandomLinkSplit`.

vijaydwivedi75 replied:
Renamed `edge_index_labeled` to `edge_label_index`.

EdisonLeeeee commented:
A gentle ping @rusty1s :)

vijaydwivedi75 commented:
Done, added dataset info in docstring!

rusty1s commented on torch_geometric/datasets/lrgb.py:
```diff
@@ -0,0 +1,294 @@
+              - 15,535
+              - ~150.94
+              - ~307.30
+              - 11
+
+    """
```
```suggestion
              - 11
    """
```

rusty1s commented on torch_geometric/datasets/lrgb.py:
```diff
@@ -0,0 +1,294 @@
+    def __init__(self, root: str, name: str, split: str = "train",
+                 transform: Optional[Callable] = None,
+                 pre_transform: Optional[Callable] = None,
+                 pre_filter: Optional[Callable] = None):
```
```suggestion
                 pre_filter: Optional[Callable] = None, ):
```

rusty1s reviewed (approved):
Thank you for adding <3
